### PR TITLE
2D Polyline

### PR DESCRIPTION
**Pull request by AimAlex**

### CODE REVIEW

**Comment by AimAlex** [comment]
Screen Shot 2019-12-05 at 8 48 02 PM
changed the color

**Comment by AimAlex** [comment]
Screen Shot 2019-12-05 at 8 52 40 PM
Screen Shot 2019-12-05 at 8 53 13 PM
reshape for linked labels

File: app/src/js/drawable/2d/polygon2d.ts [hunk 1]
```diff
@@ -520,21 +274,21 @@ export class Polygon2D extends Label2D {
     if (this.editing === true &&
       this._state === Polygon2DState.DRAW) {
       // add vertex
-      const isClosed = this.addVertex(coord)
-      if (isClosed) {
-        // finish adding when it is closed
-        this._state = Polygon2DState.CLOSED
+      const isFINISHED = this.addVertex(coord)
+      if (isFINISHED) {
```
Comment (gycn) [inline app/src/js/drawable/2d/polygon2d.ts]: Could you rename to `isFinished`?

File: app/src/js/drawable/2d/polygon2d.ts [hunk 2]
```diff
@@ -642,20 +456,301 @@ export class Polygon2D extends Label2D {
...
+    switch (point.type) {
+      case PointType.MID: // from midpoint to curve
+        const prevPoint =
+          this._points[this.getPreviousIndex(highlightedHandleIndex)]
+        const nextPoint =
+          this._points[this.getNextIndex(highlightedHandleIndex)]
+        const controlPoints = this.getCurvePoints(prevPoint, nextPoint)
+        this._points[highlightedHandleIndex] = controlPoints[0]
+        this._points.splice(
+          highlightedHandleIndex + 1, 0, controlPoints[1]
```
Comment (gycn) [inline app/src/js/drawable/2d/polygon2d.ts]: Cna you use get next index and get previous index here?

File: app/src/js/drawable/2d/polygon2d.ts [hunk 3]
```diff
@@ -642,20 +456,301 @@ export class Polygon2D extends Label2D {
...
+        )
+        break
+      case PointType.CURVE: // from curve to midpoint
+        const newMidPointIndex =
+          (this._points[highlightedHandleIndex - 1].type === PointType.CURVE) ?
+            (highlightedHandleIndex - 1 + numPoints) % numPoints :
```
Comment (gycn) [inline app/src/js/drawable/2d/polygon2d.ts]: Cna you use get next index and get previous index here?

File: app/src/js/drawable/2d/polygon2d.ts [hunk 4]
```diff
@@ -598,8 +410,10 @@ export class Polygon2D extends Label2D {
     this._state = Polygon2DState.DRAW
     const itemIndex = state.user.select.item
     this._order = state.task.status.maxOrder + 1
+    const shapeType = this._closed ?
+                LabelTypeName.POLYGON_CLOSED_2D : LabelTypeName.POLYGON_OPEN_2D
     this._label = makeLabel({
-      type: LabelTypeName.POLYGON_2D, id: -1, item: itemIndex,
+      type: shapeType, id: -1, item: itemIndex,
```
Comment (gycn) [inline app/src/js/drawable/2d/polygon2d.ts]: Can you rename `shapeType` to `labelType`?

File: app/src/js/drawable/2d/label2d_handler.ts [hunk 5]
```diff
@@ -225,6 +233,16 @@ export class Label2DHandler {
     return Session.label2dList.selectedLabels.length === 0
   }
 
+  /** returns whether selectedLabels is editing */
+  private isSelectedLabelsEditing (): boolean {
```
Comment (fyu) [inline app/src/js/drawable/2d/label2d_handler.ts]: better to be named to `isEditingSelectedLabels`

File: app/src/js/server/types.ts [hunk 6]
```diff
@@ -142,8 +142,8 @@ export const defaultBoxCategories = [
   'traffic light'
 ]
 
-/* default categories when file is missing and label is polyline2d */
-export const defaultPolyline2DCategories = [
+/* default categories when file is missing and label is polygonopen2d */
+export const defaultPolygonOpen2DCategories = [
```
Comment (fyu) [inline app/src/js/server/types.ts]: Also needs to change the name here

**Pull request merged**